### ISSUE TITLE
LocalCopyPropagation pass can cause exponential increase in expression/code size

### ISSUE THREAD

**Issue by jafingerhut** [p4lang/p4c#2316]
I have seen something similar happen 3 years ago when trying out some P4_16 code to calculate IPv4 header checksums, so I do not believe this is a recent issue -- it is probably just the way that LocalCopyPropagation has always worked, and most P4 programs do not exercise it in a way that is very noticeable to end users.

In the attached P4_16 program, there is an action with multiple calls to another action `do_round`.  If you run p4c with options like `--top4 Front,Mid` that cause it to write out intermediate P4_16 programs after each compiler pass, you will see that every time you add another sequential call to `do_round`, it roughly doubles the size of the file produced as output from the LocalCopyPropagation pass.

It seems that LocalCopyPropagation could perhaps be described as the opposite of common subexpression elimination -- it is causing common subexpressions to be duplicated in its output.  In a program like the attached one, that can lead to an exponential number of copies of these subexpressions being produced.

[proliferation1.p4.txt](https://github.com/p4lang/p4c/files/4478275/proliferation1.p4.txt)

**Comment by jafingerhut** [comment]
Note: I do not believe there is anything magic about the definition of the `do_round` action in the attached example program `proliferation1.p4.txt`.  I believe that you could replace it with just about any expression that has two or more occurrences of the parameter `x` on the right-hand side, and if the compiler cannot simplify away those 2 or more occurrences before the LocalCopyPropagation pass down to 1, then the exponential blow-up will occur.

**Comment by jafingerhut** [comment]
I understand why the exponential blow-up in code size is occurring.

I think it is reasonable to ask: Should there at least be an option to the compiler _not_ to do this?  The LocalCopyPropagation transformations are not the only way to produce correct output from the compiler -- it is a design choice which has advantages in some situations, and disadvantages in others.

**Comment by milicaperic** [comment]
There is an compiler option --excludeMidendPasses which exclude pass(es) from midend passes. 
In this case that is LocalCopyPropagation but unfortunately exponential problem is staying.

**Comment by mbudiu-vmw** [comment]
Can you make sure that the pass is not run by running the compiler with the -v flag, which lists all executed passes?

**Comment by milicaperic** [comment]
You are right. 
When I use excludeMidendPasses option for bmv backend there is no exponential problem, 
but when I used p4test excludeMidendPasses option the problem remains.
So, I looked at the code and noticed that excludeMidendPasses option is not processed  in p4test as in bmv2, 
although it says in --help that it exists.
Should I add it to work for p4test backend as well?

**Comment by mbudiu-vmw** [comment]
That would be very useful.

**Comment by milicaperic** [comment]
Should we close this issue since there is an option to exclude pass and solve exponential problem?

**Comment by jafingerhut** [comment]
Regarding closing this issue, is there no other reason to run the LocalCopyPropagation pass other than the behavior that leads to exponential expression size blowup in the provided test program?  Is that pass's entire purpose to take expressions assigned to variables, and 'push' those expressions later in the program wherever the variable is read?

**Comment by mbudiu-vmw** [comment]
No, I think this is still a bug. The solution would be for the compiler to use some policy and avoid replacing expressions that are too complex.
